include/irr/asset/CTexturePacker.h

```diff
@@ -53,20 +55,119 @@ class CSquareAddressAllocator
     _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_MORTON_MASK = (1u<<ADDR_LAYER_SHIFT)-1u;
 
-    using addr_alctr_t = core::GeneralpurposeAddressAllocator<uint32_t>;
+    const uint32_t m_pgSzxy;
+    const uint32_t m_tilesPerDim;
+    CSquareAddressAllocator<uint32_t> m_addrAlctr;
+
+    core::smart_refctd_ptr<ICPUImage> m_pageTable;
+public:
+    struct STexOffset
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t X_MASK = ((1u<<(ADDR_LAYER_SHIFT>>1))-1u);
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_addr = (~0u);
+
+        STexOffset(uint32_t _addr) : addr(_addr) {}
+        STexOffset(uint32_t _x, uint32_t _y, uint32_t _layer)
+        {
+            addr = (_x&X_MASK);
+            addr |= (_y&X_MASK)<<(ADDR_LAYER_SHIFT>>1);
+            addr |= (_layer<<ADDR_LAYER_SHIFT);
+        }
+        uint32_t x() const { return addr&X_MASK; }
+        uint32_t y() const { return (addr>>(ADDR_LAYER_SHIFT>>1))&X_MASK; }
+        uint32_t layer() const { return addr>>ADDR_LAYER_SHIFT; }
+
+        uint32_t addr;
+    };
+
+    ITexturePacker(uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u) :
+        m_pgSzxy(_pgSzxy),
+        m_tilesPerDim(_tilesPerDim),
+        m_addrAlctr(14u, _tilesPerDim, _numLayers)
+    {
+        ICPUImage::SCreationParams params;
+        params.arrayLayers = _numLayers;
+        params.extent = {_tilesPerDim,_tilesPerDim,1u};
+        params.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0);
+        params.format = EF_R32_UINT;
+        params.mipLevels = core::findMSB(_tilesPerDim)+1;
+        params.samples = ICPUImage::ESCF_1_BIT;
+        params.type = IImage::ET_2D;
+        m_pageTable = ICPUImage::create(std::move(params));
+        
+        auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(m_pageTable->getCreationParameters().mipLevels);
+
+        uint32_t bufOffset = 0u;
+        for (uint32_t i = 0u; i < m_pageTable->getCreationParameters().mipLevels; ++i)
+        {
+            const uint32_t tilesPerLodDim = _tilesPerDim>>i;
+            const uint32_t regionSz = tilesPerLodDim*tilesPerLodDim*_numLayers*4u;
+            auto& region = (*regions)[i];
+            region.bufferOffset = bufOffset;
+            region.bufferImageHeight = 0u;
+            region.bufferRowLength = tilesPerLodDim*4u;
+            region.imageExtent = {tilesPerLodDim,tilesPerLodDim,1u};
+            region.imageOffset = {0,0,0};
+            region.imageSubresource.baseArrayLayer = 0u;
+            region.imageSubresource.layerCount = _numLayers;
+            region.imageSubresource.mipLevel = i;
+            region.imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+
+            bufOffset += regionSz;
+        }
+        auto buf = core::make_smart_refctd_ptr<ICPUBuffer>(bufOffset);
+
+        m_pageTable->setBufferAndRegions(std::move(buf), regions);
```

Comment:
only ICPUTexturePacker needs to have the page table in CPU memory
(GPU version will most likely work, by queuing up the allocations and deallocations in a dynamic `core::vector` and passing that as an SSBO toa compute shader that will update the pages on a GPU-resident-only page table texture)